AC_Autotune: Incorporate autotune for Tradheli

bnsgeyer opened the pull request:
This PR reworks the Autotune library to include an autotune feature for traditional helicopters.  I have started a PR just to get visibility to the developers.  This is a significant undertaking as I had to divide the library into subclasses for Multi and Heli.  I had to divide the library because I had to take a completely different approach for autotune for heli.  I am keeping the sequence of events of waiting for level, testing, and update gains.  But I am changing the tuning sequence as well as adding additional tune types like feedforward gain tuning.  Also increasing the library of available test inputs to include doublets, dwells and maybe even chirps.

To date, I have split the library and coded up the Feedforward tuning and Rate P up.  Here is a link to post on discuss that demos what i've done to date.  
https://discuss.ardupilot.org/t/is-autotune-for-traditional-helicopter-in-sight/54619/11

Addresses issue #3882

andyp1per commented on libraries/AC_AutoTune/AC_AutoTune.cpp:
```diff
@@ -675,102 +469,36 @@ void AC_AutoTune::control_attitude()
-                updating_angle_p_up(tune_yaw_sp, AUTOTUNE_SP_MAX, AUTOTUNE_SP_STEP, target_angle, test_angle_max, test_rate_min, test_rate_max);
-                break;
-            }
+            updating_angle_p_down_all(axis);
```
updating_angle_p_up_all()?

bnsgeyer replied:
Wasn't sure how to do this.  the tradheli version of updating_ methods uses different arguments.  And they are called from a common method in the autotune library.  I know that I can have separate versions of the method containing separate arguments but then I would have to know which frame type it is to know which method to use.

andyp1per commented on libraries/AC_AutoTune/AC_AutoTune.cpp:
```diff
@@ -255,8 +205,8 @@ void AC_AutoTune::send_step_string()
         return;
-    case TWITCHING:
-        gcs().send_text(MAV_SEVERITY_INFO, "AutoTune: TWITCHING");
+    case TESTING:
```
I'm not a fan of this name change. Testing for me is the pilot testing afterwards.

bnsgeyer replied:
This change was in the subclass split PR which @rmackay9 reviewed.  I didn't receive any comments from him on this.  I went through significant testing of the subclass split PR to show that the multi autotune operated nearly identical to autotune before the split.

bnsgeyer commented:
@andyp1per  if you have time in an hour or so, I like to chat with you about this PR and your comments.

andyp1per commented:
@bnsgeyer ping me on discord  - if I am around I will answer!

bnsgeyer commented:
@tridge I have reset this branch to my latest branch that includes all of heli autotune code.  Please have a look and direct me as to how to handle the #defines when they are moved to the cpp files.  I have moved them all over to the cpp files but now am having difficulty trying decide how to structure them.  It seems that some would have to be defined in the subclass cpp because there are methods that use them but are also used in the common methods in the parent class.  That would mean future changes would require the dev to make the changes in each sub class even though the #define may be common to both subclasses.
  Thanks in advance!

bnsgeyer commented:
@tridge I have retained the minimum #defines used by both sub classes within Autotune.cpp only.  Those that are used by autotune.cpp and the sub class, I moved to the subclass and created an accessor in the subclass for autotune.cpp.  The issue I am having now is that I get a bunch of warnings that I have redefined #defines.  Hopefully you can help me out with this and get me on the right track. Thanks!

tridge commented:
@bnsgeyer I've put a fix for the include issue here:
https://github.com/tridge/ardupilot/commits/pr-autotune

bnsgeyer commented:
@rmackay9  and @peterbarker I am to the point that I would like to have you start picking away at this PR.  Ultimately the last two commits will be removed before merging.  For now they help test the code.  I know there is a lot of small stuff to start tidying up.  I have some unused methods to remove.   Ultimately I want to keep as much generic methods in the autotune.cpp file so at some point we can add heli autotune as an option for multi's to run.  Also I have asked @Gone4Dirt to add a BUILD parameter for arduheli that I plan to use in this PR once it hits master.  Lastly, I have added parameters and pulled the Multi specific parameter into the multi class but it causes some issues when upgrading.  Any ideas on how to do that better would be great.
Thanks in advance!

peterbarker commented on libraries/AC_AutoTune/AC_AutoTune.cpp:
```diff
@@ -788,56 +570,89 @@ void AC_AutoTune::control_attitude()
+                break;
+            }
+
+            // increment the tune type to the next one in tune sequence
+            tune_seq_curr++;
+            tune_type = tune_seq[tune_seq_curr];
```
I think we might be better off with
```suggestion
            // move to next tune type in tune sequence
            tune_type = next_tune_type(tune_type);
```

where `next_tune_type` is

```
virtual TuneType next_tune_type(TuneType t) const = 0;
```

on `AC_AutoTune`

That method would be a simple `switch` on Multi (potentially just the dodgy `+1` crap we already use in `master`, however).  The Heli one looks more complicated based on parameters :-)

bnsgeyer replied:
There was a lot of thought that went behind this.  The idea was to allow the user to specify the sequence of tests they want to use whether it was just the FF tuning, Rate D and Rate P, Angle P or all of them.  However we decide to proceed, I want heli to be able to choose the sequence of tuning tests.

bnsgeyer replied:
Added a method as requested but wasn't able to do it as a function that returned a TuneType.  So just used the argument that returned the tune type.

peterbarker commented on libraries/AC_AutoTune/AC_AutoTune.cpp:
```diff
@@ -1713,56 +1290,831 @@ void AC_AutoTune::get_poshold_attitude(float &roll_cd_out, float &pitch_cd_out,
+                attitude_control->input_angle_step_bf_roll_pitch_yaw(0.0f, dir_sign * target_angle, 0.0f);
+                break;
+            case YAW:
+                // request pitch to 20deg
```
```suggestion
                // request yaw to movement
```

peterbarker commented on libraries/AC_AutoTune/AC_AutoTune.cpp:
```diff
@@ -1713,56 +1290,831 @@ void AC_AutoTune::get_poshold_attitude(float &roll_cd_out, float &pitch_cd_out,
+            // Testing increasing stabilize P gain so will set lean angle target
+            switch (test_axis) {
+            case ROLL:
+                // request roll to 20deg
```
Pretty sure the 20 here is wrong

bnsgeyer replied:
What do you mean it's wrong?  This is the same code copied from above but put into a method.  This is straight from the original autotune.  we'll have to have @lthall check it.

peterbarker commented on libraries/AC_AutoTune/AC_AutoTune.cpp:
```diff
@@ -1713,56 +1290,831 @@ void AC_AutoTune::get_poshold_attitude(float &roll_cd_out, float &pitch_cd_out,
+        }
+    }
+
+    // capture this iterations rotation rate and lean angle
```
```suggestion
    // capture this iteration's rotation rate and lean angle
```

bnsgeyer replied:
done.  thanks!

peterbarker commented on libraries/AC_AutoTune/AC_AutoTune.cpp:
```diff
@@ -1713,56 +1290,831 @@ void AC_AutoTune::get_poshold_attitude(float &roll_cd_out, float &pitch_cd_out,
+    float command_reading = 0.0f;
+    float tgt_rate_reading = 0.0f;
+    const uint32_t now = AP_HAL::millis();
+    static float trim_command_reading = 0.0f;
+    static float trim_heading = 0.0f;
+    static float rate_request_cds;
+    static float angle_request_cd;
```
member variables instead

bnsgeyer replied:
done

peterbarker commented on libraries/AC_AutoTune/AC_AutoTune.cpp:
```diff
@@ -1713,56 +1290,831 @@ void AC_AutoTune::get_poshold_attitude(float &roll_cd_out, float &pitch_cd_out,
+    static float rate_request_cds;
+    static float angle_request_cd;
+
+    // TODO make filter leak dependent on dt
```
Does this need to be done before this is merged?

bnsgeyer replied:
done

peterbarker reviewed (changes requested):
Entry from GUIDED isn't permitted - perhaps it should be?

Parameter conversion may be required for the parameters you're moving to the subclass.  That shouldn't cause tests to fail, however.

Perhaps you could leave the parameters in the base class for the time being, even if they're not relevant for heli.

What do you think about factoring some of this out earlier.  If we've got general agreement on the approach you've taken here, it should be possible to a chunk of this as "obviously OK" patches, like factoring the logging out.  Actually creating the MultiCopter subclass ahead of time would also be possible - since a lot of that is moving unmodified code, it can be merged relatively easily.

I've only done a partial review here - its a massive set of changes which is going to take a bit of getting my head around :-)

bnsgeyer commented:
@peterbarker i had gone through making a PR that was for moving heli and multi into subclasses here PR #15705. After @rmackay9 had conducted a thorough review, he decided not to merge it and that I should just continue with this PR and build it on the commits of the other one.  If you would like I can squash all my commits since that point and that should make reviewing the delta easier

Gone4Dirt commented on ArduCopter/wscript:
```diff
@@ -45,5 +45,5 @@ def build(bld):
         program_groups=['bin', 'heli'],
         use=vehicle + '_libs',
-        defines=['FRAME_CONFIG=HELI_FRAME'],
+        defines=['FRAME_CONFIG=HELI_FRAME', 'HELI_BUILD'],
```
I have done this PR #18784 as first attempt to resolve this.

bnsgeyer commented on libraries/AC_AutoTune/AC_AutoTune.cpp:
```diff
@@ -390,7 +332,7 @@ void AC_AutoTune::run()
         }
     }
-    if (zero_rp_input) {
+    if (zero_rp_input && !allow_pilot_rp_input()) {
```
This allows the pilot to make inputs while the automated testing is being conducted.  Some of the tests are 20 seconds long and the pilot is allowed to make small inputs to help maintain position during the automated Angle P test.  The multicopter autotune was designed to automatically make inputs in between twitches to help maintain position, if autotune was entered from a GPS aided mode.  Mine is different because it is doing it during the automated test.  Now that I have shown that the rate P and rate D tests are being held by velocity feedback, I may remove the pilot input and implement the same code for the angle P test.

bnsgeyer replied:
Removed pilot input during ANGLE_P tuning and added velocity feedback.

bnsgeyer commented on libraries/AC_AutoTune/AC_AutoTune.cpp:
```diff
@@ -90,15 +67,20 @@
-#define AUTOTUNE_MESSAGE_TESTING 5
-
-#define AUTOTUNE_ANNOUNCE_INTERVAL_MS 2000
+// ifdef is not working.  Modified multi values to reflect heli requirements
```
PR was merged and I have adjusted this to use the new APM_BUILD_Heli

bnsgeyer commented on libraries/AC_AutoTune/AC_AutoTune.h:
```diff
@@ -220,9 +261,75 @@ class AC_AutoTune {
+    float    command_out;
+    float    test_tgt_rate_filt;                    // filtered target rate
+    float    filt_target_rate;
+    bool     ff_up_first_iter   : 1;       //true on first iteration of ff up testing
```
Rebased on master.  then fixed this.

bnsgeyer commented on libraries/AC_AutoTune/AC_AutoTune.cpp:
```diff
@@ -110,19 +92,7 @@ const AP_Param::GroupInfo AC_AutoTune::var_info[] = {
-    // @Range: 0.001 0.006
-    // @User: Standard
-    AP_GROUPINFO("MIN_D", 3, AC_AutoTune, min_d,  0.001f),
+// Indices 2 and 3 where AGGR and MIN_D.  These were moved to the Multi SubClass
```
parameters were moved to subclasses.  So this comment is no longer valid

bnsgeyer commented on libraries/AC_AutoTune/AC_AutoTune.cpp:
```diff
@@ -171,6 +147,16 @@ bool AC_AutoTune::init_internals(bool _use_poshold,
         // we are restarting tuning so restart where we left off
+        // reset test variables to continue where we left off
+        // reset dwell test variables if sweep was interrupted in order to restart sweep
+        if (!is_equal(start_freq,stop_freq)) {
```
done. thanks!

bnsgeyer commented on libraries/AC_AutoTune/AC_AutoTune.cpp:
```diff
@@ -506,169 +440,64 @@ void AC_AutoTune::control_attitude()
         }
 
-        float target_max_rate;
+        // Initialize test specific variables
```
done. Thanks!

rmackay9 commented on libraries/AC_AutoTune/AC_AutoTune_FreqResp.cpp:
```diff
@@ -0,0 +1,448 @@
+        float tgt_ampl = 0.0f;
+        uint32_t meas_time = 0;
+        uint32_t tgt_time = 0;
+        for (int i = 0;  i < AUTOTUNE_DWELL_CYCLES; i++) {
```
"int" should probably "uint8_t" ..

bnsgeyer replied:
done

rmackay9 commented on libraries/AC_AutoTune/AC_AutoTune_FreqResp.h:
```diff
@@ -0,0 +1,87 @@
+    float get_accel_max() { return max_accel; }
+
+private:
+    float max_target, max_meas, prev_target, prev_meas, prev_tgt_angle, prev_meas_angle;
```
Just a formatting things but I wonder if we should perhaps separate these variable declarations onto separate lines and put comments beside each as to it's purpose.  This would be more consistent with other libraries.

bnsgeyer replied:
done.

rmackay9 commented on libraries/AC_AutoTune/AC_AutoTune_FreqResp.h:
```diff
@@ -0,0 +1,87 @@
+    bool cycle_complete = false;
+    float curr_test_freq, curr_test_gain, curr_test_phase;
+    float max_accel;
+    InputType excitation;
```
Let's rename "excitation" to something that is more easily understood.  For me, "excitation" is more like an adjective than a noun.

Gone4Dirt replied:
Excitation is the correct technical term.  I vote it stays the same.

rmackay9 replied:
@Gone4Dirt I agree in general that AP should use pre-defined terminology where possible but is "excitation" actually used as a noun meaning "InputType"?  A quick google search turns up this definition, "the application of energy to something"... that sounds more like a verb to me.  I'm neither a controls expert nor an english major but it seems more like a "what are we doing?" rather than a "what is it?".

rmackay9 replied:
ok, excitation is OK with me..

rmackay9 commented on libraries/AC_AutoTune/AC_AutoTune_FreqResp.h:
```diff
@@ -0,0 +1,87 @@
+    };
+    ObjectBuffer<peak_info> *meas_peak_info_buffer;
+    ObjectBuffer<peak_info> *tgt_peak_info_buffer;
+    void push_to_meas_buffer(uint16_t count, float amplitude, uint32_t time_ms);
```
comments above these method declarations would be helpful for future developers I think.

bnsgeyer replied:
Done.

rmackay9 commented on libraries/AC_AutoTune/AC_AutoTune_Heli.cpp:
```diff
@@ -0,0 +1,1200 @@
+    AP_Param::setup_object_defaults(this, var_info);
+}
+
+void AC_AutoTune_Heli::test_init()
```
maybe add a small comment above these methods.  The others all have them so it might help a bit.

bnsgeyer replied:
Done.

rmackay9 commented on libraries/AC_AutoTune/AC_AutoTune_Heli.cpp:
```diff
@@ -0,0 +1,1200 @@
+    // @Description: Defines the start frequency for sweeps and dwells
+    // @Range: 10 30
+    // @User: Standard
+    AP_GROUPINFO("MIN_FRQ", 3, AC_AutoTune_Heli, min_sweep_freq,  10.0f),
```
I'd prefer if the "MIN" or "MAX" went at the end of the variable name instead of in the middle.

bnsgeyer replied:
done

rmackay9 commented on libraries/AC_AutoTune/AC_AutoTune_Heli.cpp:
```diff
@@ -0,0 +1,1200 @@
+            curr_test_freq = freq[freq_cnt];
+            sp_prev_gain = gain[freq_cnt];
+        } else if (gain[freq_cnt] > 1.1f * max_resp_gain && tune_p > AUTOTUNE_SP_MIN && !find_peak) {
+                tune_p -= gain_incr;
```
small white space issue here I think..

bnsgeyer replied:
done

rmackay9 commented on libraries/AC_AutoTune/AC_AutoTune_Multi.cpp:
```diff
@@ -0,0 +1,673 @@
+            ignore_next = true;
+            // bounce back is bigger than our threshold so increment the success counter
+            counter++;
+        } else {
```
nitpick: I think this line and the line below can be combined into a single "else if" and "ignore_next == false" could be replaced with "!ignore_next".

bnsgeyer replied:
it was existing code.

rmackay9 commented on libraries/AC_AutoTune/AC_AutoTune.h:
```diff
@@ -16,14 +16,32 @@
+#define AUTOTUNE_SUCCESS_COUNT                4     // The number of successful iterations we need to freeze at current gains
+
+// Auto Tune message ids for ground station
+#define AUTOTUNE_MESSAGE_STARTED 0
```
I think maybe AUTOTUNE_MESSAGE_STARTED (and similar) are only used in AC_AutoTune.cpp so they remain back in the .cpp file.

bnsgeyer replied:
@rmackay All of these messages except AUTOTUNE_MESSAGE_SAVED_GAINS are used by methods only in the parent class.   AUTOTUNE_MESSAGE_SAVED_GAINS is used in the sub classes.  So I was thinking of keeping all of them together and in the AutoTune.h file.

rmackay9 commented on libraries/AC_AutoTune/AC_AutoTune.cpp:
```diff
@@ -1728,56 +1282,824 @@ void AC_AutoTune::get_poshold_attitude(float &roll_cd_out, float &pitch_cd_out,
+                    AP::scheduler().get_loop_period_s());
+    command_out = command_filt.apply((command_reading - trim_command_reading),
+                                     AP::scheduler().get_loop_period_s());
+    filt_target_rate = target_rate_filt.apply(tgt_rate_reading,
```
This line is producing a compile error for me at least.
```
[752/818] Compiling ArduCopter/Copter.cpp
../../libraries/AC_AutoTune/AC_AutoTune.cpp: In member function 'void AC_AutoTune::rate_ff_test_run(float, float, float)':
../../libraries/AC_AutoTune/AC_AutoTune.cpp:1567:46: error: 'tgt_rate_reading' may be used uninitialized in this function [-Werror=maybe-uninitialized]
     filt_target_rate = target_rate_filt.apply(tgt_rate_reading,
                        ~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~
                        AP::scheduler().get_loop_period_s());
                        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated due to -Wfatal-errors.
```

bnsgeyer replied:
@rmackay9  I was recently working on removing some static variables and hadn’t completed the work.  Remove the last commit and it should compile and run fine.

rmackay9 commented on libraries/AC_AutoTune/AC_AutoTune.h:
```diff
@@ -67,44 +92,121 @@ class AC_AutoTune {
     bool init_position_controller();
 
-private:
+    // main state machine to level vehicle, perform a test and update gains
```
I think that "control_attitude()" can remain "private:" (as it was before) because it is only called from within AC_AutoTune.cpp.

bnsgeyer replied:
made all methods only used in autotune.cpp private

rmackay9 commented on libraries/AC_AutoTune/AC_AutoTune.h:
```diff
@@ -215,14 +324,120 @@ class AC_AutoTune {
     AP_Int8  axis_bitmask;
     AP_Float aggressiveness;
     AP_Float min_d;
+    AP_Float vel_hold_gain;
```
It's odd that these parameters are declared in the parent class, AC_AutoTune,h but this class doesn't have a var_table.  We should probably discuss with @peterbarker but another option would be to leave AC_AutoTune with a var_table and then also allow the child classes to have var tables.  I'm sure we do this in other classes.

rmackay9 replied:
I'll help with this.

bnsgeyer replied:
@rmackay9 moved code that changed gain values based on #defines to the subclasses.  I had used helper functions to get specific heli and multi defines but felt it better to put this code into the subclasses.  This moved all but one (axis_bitmask) parameter variable into the subclasses.  So I put those variable declarations into the subclass.  For axis_bitmask, I made an accessor so the parent class can access it and subsequently moved its declaration to the subclass.  Hope this makes sense and helps with your concern here.

rmackay9 commented on libraries/AC_AutoTune/AC_AutoTune.h:
```diff
@@ -215,14 +324,120 @@ class AC_AutoTune {
+    void dwell_test_run(uint8_t freq_resp_input, float start_frq, float stop_frq, float &dwell_gain, float &dwell_phase);
+
+    // dwell test used to perform frequency dwells for angle gains
+    void angle_dwell_test_init(float filt_freq);
```
It seems like "angle_dwell_test_init" is not called from anywhere?

I think maybe the dwell test related methods and variables should be moved to the AC_AutoTune_Heli class because they aren't used by _Multi.

bnsgeyer replied:
good catch.  I'm surprised that didn't affect my angle dwell test but maybe they are very similar because they mostly use the same variables.  I spent some time cleaning up the dwell test methods but it was too much to try and consolidate into one method, at least right now.  

We should have a discussion on moving the multi and heli specific tests into the respective classes.  I have my reasons not to but you may not agree with them.

rmackay9 commented on libraries/AC_AutoTune/AC_AutoTune.h:
```diff
@@ -215,14 +324,120 @@ class AC_AutoTune {
+    float    filt_target_rate;
+    bool     ff_up_first_iter;                      //true on first iteration of ff up testing
+    float    test_gain[20];                             // gain of output to input
+    float    test_freq[20];
```
it would be good to put comments beside all these variables to explain their purpose.

bnsgeyer replied:
done

rmackay9 commented on libraries/AC_AutoTune/AC_AutoTune.h:
```diff
@@ -215,14 +324,120 @@ class AC_AutoTune {
+    float    trim_pff_out;  // trim output of the PID rate controller for P, I and FF terms
+    float    trim_meas_rate;  // trim measured gyro rate
+
+    //variables from rate FF test
```
similar to the dwell test I think we should move the rate FF test related methods and variables to the Heli class.

bnsgeyer replied:
done

rmackay9 commented on libraries/AC_AutoTune/AC_AutoTune_FreqResp.h:
```diff
@@ -0,0 +1,87 @@
+    AC_AutoTune_FreqResp()
+{
+    // initialize test variables
+    meas_peak_info_buffer = new ObjectBuffer<peak_info>(AUTOTUNE_DWELL_CYCLES);
```
not a big deal of course but I'd probably move the implementation of the constructor to the .cpp file.

bnsgeyer replied:
Can't move out of the header as the heli subclass uses the define.  Should I make an accessor for the subclass since this is a standalone.  I guess it seems odd that the subclass is using the define directly

rmackay9 replied:
ok, if the heli subclass uses it then leaving it in the .h seems OK to me.

rmackay9 commented on libraries/AC_AutoTune/AC_AutoTune_FreqResp.h:
```diff
@@ -0,0 +1,87 @@
+        uint16_t curr_count;
+        float amplitude;
+        uint32_t time_ms; 
+
```
nitpick: extra blank line here.

bnsgeyer replied:
done.

rmackay9 commented on libraries/AC_AutoTune/AC_AutoTune_FreqResp.cpp:
```diff
@@ -0,0 +1,448 @@
+    max_accel = 0.0f;
+    max_meas_rate = 0.0f;
+    max_command = 0.0f;
+    meas_peak_info_buffer->clear();
```
Theoretically at least the allocation of meas_peak_info_buffer could have failed so we might want to have a nullptr check here.  How that failure is handled more broadly is perhaps a more difficult question to answer though.

bnsgeyer replied:
fixed.

rmackay9 commented on libraries/AC_AutoTune/AC_AutoTune_FreqResp.cpp:
```diff
@@ -0,0 +1,448 @@
+        float tgt_ampl = 0.0f;
+        uint32_t meas_time = 0;
+        uint32_t tgt_time = 0;
+        for (int i = 0;  i < AUTOTUNE_DWELL_CYCLES; i++) {
```
this "int" should probably be "uint8_t".

bnsgeyer replied:
done

rmackay9 commented on libraries/AC_AutoTune/AC_AutoTune.cpp:
```diff
@@ -160,6 +117,12 @@ bool AC_AutoTune::init_internals(bool _use_poshold,
 
     case UNINITIALISED:
+        // initializes dwell test sequence for rate_p_up and rate_d_up tests for tradheli
+        freq_cnt = 0;
```
I wonder if this initialisation should instead be put into a heli specific version of "backup_gains_and_initialise".  So I'm thinking heli implements backup_gains_and_initialise() and within it's implementation it does this extra stuff and calls the parent AC_AutoTune::backup_gains_and_initialise().

bnsgeyer replied:
done

rmackay9 commented on libraries/AC_AutoTune/AC_AutoTune_Multi.cpp:
```diff
@@ -0,0 +1,673 @@
+        gcs().send_text(MAV_SEVERITY_INFO, "AutoTune: p=%f d=%f", (double)tune_rp, (double)tune_rd);
+        break;
+    case RFF_UP:
+    case RFF_DOWN:
```
maybe we should record a flow_of_control internal error if we get to this state or MAX_GAINS or TUNE_COMPLETE.

bnsgeyer replied:
done.  Just for Max_Gains.  Tune_complete is used by multi

rmackay9 commented on libraries/AC_AutoTune/AC_AutoTune.cpp:
```diff
@@ -1728,56 +1282,824 @@ void AC_AutoTune::get_poshold_attitude(float &roll_cd_out, float &pitch_cd_out,
-
-// Write an Autotune data packet
-void AC_AutoTune::Log_Write_AutoTune(uint8_t _axis, uint8_t tune_step, float meas_target, float meas_min, float meas_max, float new_gain_rp, float new_gain_rd, float new_gain_sp, float new_ddt)
+void AC_AutoTune::twitch_test_init()
```
maybe twitch_test_init is multicopter specific and should be moved to AC_AutoTune_Multi?

bnsgeyer replied:
done

rmackay9 commented on libraries/AC_AutoTune/AC_AutoTune.cpp:
```diff
@@ -506,169 +452,69 @@ void AC_AutoTune::control_attitude()
         }
 
+        // protect from roll over
+        if (ahrs_view->roll_sensor > 3000.0f || ahrs_view->roll_sensor < -3000.0f ||ahrs_view->pitch_sensor > 3000.0f || ahrs_view->pitch_sensor < -3000.0f) {
```
The maximum lean angle (3000) should probably moved to a definition somewhere.   I agree with @peterbarker's comments that an overall lean angle test might be better.

bnsgeyer replied:
fixed

rmackay9 commented on libraries/AC_AutoTune/AC_AutoTune.cpp:
```diff
@@ -1022,31 +825,31 @@ void AC_AutoTune::load_tuned_gains()
+            attitude_control->get_rate_roll_pid().kI(get_tuned_ri(axis));
             attitude_control->get_rate_roll_pid().kD(tune_roll_rd);
-            attitude_control->get_rate_roll_pid().ff(orig_roll_rff);
+            attitude_control->get_rate_roll_pid().ff(tune_roll_rff);
```
I guess this is a small change in behaviour that we should check with @lthall.  Basically it means that if a multicopter users had set the ATC_RAT_RLL_FF to a non-zero value, after a successful (or unsuccessful?) autotune it would be set back to zero.

bnsgeyer replied:
Not necessarily. In the initialize gains method, I added a tune_roll_rff that is set to the same value as the orig_roll_rff gain.  So this should not change the FF gain after the vehicle is tuned.

bnsgeyer replied:
load and backup gain methods moved to subclasses

rmackay9 commented:
@bnsgeyer, I've made some fairly superficial suggestions above.  I'm still going through the PR to see if I can spot any issues with the change in structure.

rmackay9 commented on libraries/AC_AutoTune/AC_AutoTune.cpp:
```diff
@@ -1063,65 +866,66 @@ void AC_AutoTune::load_intra_test_gains()
         attitude_control->get_rate_roll_pid().filt_T_hz(orig_roll_fltt);
         attitude_control->get_rate_roll_pid().slew_limit(orig_roll_smax);
         attitude_control->get_angle_roll_p().kP(orig_roll_sp);
+        attitude_control->set_accel_roll_max_cdss(orig_roll_accel);
```
So we have as slight change in behaviour but I think it is an improvement.
AC_AutoTune::load_intra_test_gains() is setting the max acceleration to the original acceleration while in master it would be zero (meaning no max acceleration).

I can imagine that during the return to level the response could theoretically be a little slower.

FYI @lthall 

FYI @lthall

bnsgeyer replied:
moved backup and load gains to subclasses

rmackay9 commented on libraries/AC_AutoTune/AC_AutoTune.cpp:
```diff
@@ -1063,65 +866,66 @@ void AC_AutoTune::load_intra_test_gains()
+// load_test_gains - load the to-be-tested gains for a single axis
 // called by control_attitude() just before it beings testing a gain (i.e. just before it twitches)
-void AC_AutoTune::load_twitch_gains()
+void AC_AutoTune::load_test_gains()
```
I think it is perhaps too confusing to split-up load_test_gains into three methods.  So we have the base method (this one), then overrides in the _Multi and _Heli classes which call the base.  I think it's just a bit too confusing for the reader to have to move back and forth between them to figure out what's going on.  Better to just force the sub classes to implement the method completely by adding a "=0" at the end of the definition in AC_AutoTune.h.  E.g. like this:

```
    // load gains for next test.  relies on axis variable being set
    virtual void load_test_gains();
```

bnsgeyer replied:
Done

rmackay9 commented on libraries/AC_AutoTune/AC_AutoTune.cpp:
```diff
@@ -1164,15 +968,10 @@ void AC_AutoTune::save_tuning_gains()
 
     // sanity check the rate P values
-    if ((axes_completed & AUTOTUNE_AXIS_BITMASK_ROLL) && roll_enabled() && !is_zero(tune_roll_rp)) {
+    if ((axes_completed & AUTOTUNE_AXIS_BITMASK_ROLL) && roll_enabled() && (!is_zero(tune_roll_rp) || allow_zero_rate_p())) {
```
I think here too in "save_tuning_gains()" it is too confusing to have three implementations.  Let's reduce it by two by removing the base class implementation.  In particular I worry that we could easily end up with situation where half the gains are saved because of a small error/mismatch in the sanity checks.

and it would be difficult to determine why.

bnsgeyer replied:
Done

rmackay9 commented on libraries/AC_AutoTune/AC_AutoTune_Heli.h:
```diff
@@ -0,0 +1,149 @@
+    void updating_rate_p_up_all(AxisType test_axis) override;
+
+    // update gains for the rate p down tune type
+    void updating_rate_p_down_all(AxisType test_axis) override {};
```
It looks like updating_rate_p_down_all() is not implemented or called from anywhere?

bnsgeyer replied:
no.  its removed.

rmackay9 commented on libraries/AC_AutoTune/AC_AutoTune_Multi.h:
```diff
@@ -0,0 +1,134 @@
+    void updating_rate_d_down_all(AxisType test_axis) override;
+
+    // update gains for the rate ff up tune type
+    void updating_rate_ff_up_all(AxisType test_axis) override {};
```
I think it might be a good idea to add internal errors to the implementation of multicopter methods that we think should never be called.  This will add a little flash but it will also help us catch programming errors which are very possible because this is a fairly complex change.

bnsgeyer replied:
Not sure what you mean by this statement "add internal errors"

rmackay9 replied:
```
        if (!wp_nav->get_wp_destination_loc(default_loc)) {
            // this should never happen
            INTERNAL_ERROR(AP_InternalError::error_t::flow_of_control);
        }
```

bnsgeyer replied:
done

rmackay9 commented:
I guess at some point you'll need to rebase this on master because PR https://github.com/ArduPilot/ardupilot/pull/18658 has renamed "inertial_nav::get_position();" to "inertial_nav->get_position_neu_cm();"

rmackay9 commented on libraries/AC_AutoTune/AC_AutoTune_Heli.cpp:
```diff
@@ -0,0 +1,1200 @@
+#define AUTOTUNE_FFI_RATIO_FINAL           0.5f     // I is set 0.5x VFF after testing
+#define AUTOTUNE_PI_RATIO_FINAL            1.0f     // I is set 1x P after testing
+#define AUTOTUNE_YAW_PI_RATIO_FINAL        0.1f     // I is set 1x P after testing
+#define AUTOTUNE_RD_STEP                  0.0005f     // minimum increment when increasing/decreasing Rate D term
```
It would be good to put these definitions in the same order in _Heli.cpp and _Multi.cpp.

bnsgeyer replied:
done

rmackay9 commented on libraries/AC_AutoTune/AC_AutoTune_Heli.cpp:
```diff
@@ -0,0 +1,1200 @@
+    // @Description: Velocity xy P gain used to hold position during Max Gain, Rate P, and Rate D frequency sweeps
+    // @Range: 0 1
+    // @User: Standard
+    AP_GROUPINFO("VELXY_P", 6, AC_AutoTune_Heli, vel_hold_gain,  0.1f),
```
I may have mentioned it somewhere else in the review already but we should discuss the sharing of this var_table between multi and helis.  In particular if a user moves between TradHeli and Multicopter the parameters are not wiped. This means that Multicopter's "AGGR" parameter value becomes TradHeli's "SEQ".

bnsgeyer replied:
@rmackay9 the vel_hold_gain param probably wouldn’t ever be used by multicopter using the twitch test technique since the test happens so quickly and it is expected to drift during the test.  I will see with the recent changes that pulled many of the tests into subclasses whether these still need to be declared at the parent level.

bnsgeyer replied:
pulled it into the subclass as it was only used by heli and with recent changes, it's use was only in the subclass.

rmackay9 commented:
It would be really helpful if we could somehow prove that he exact same inputs result in the same outputs before and after this PR.  A sort of Replay for AP's vehicle code or at least AutoTune.  I'm not making this a hard condition for the merging of this PR but the complexity of this change makes it difficult to be sure that multicopter's autotune behaviour is the same before and after this change.

The need to jump up and down between the base class and the two child classes makes understanding AutoTune very difficult.  I can't help but think we need better names, better ordering of methods and/or commenting of the code to make it understandable to future developers who dare to try to and understand it.  The structure needs to accommodate two dimensions: the vehicle (multi, tradheli) and the various tests (Rate P, Rate D, Angle P, Rate FF).. it would be nice if we could make this 2D implementation visible and somewhat navigable by the developer.

.. so to point out one example, there is a new MAX_GAINS test which appears in the "TuneType" enum but then the "updating_max_gains_all" is only implemented for TradHeli.  I think there should be a "TradHeli only" comment beside that enum value and/or we should move the updating_max_gains_all() up to the AC_AutoTune.h but then not have multicopters implement it.. or if they do it should raise an internal error.

rmackay9 commented on libraries/AC_AutoTune/AC_AutoTune.h:
```diff
@@ -215,14 +324,120 @@ class AC_AutoTune {
+    // variables from angle dwell test
+    float trim_yaw_tgt_reading;
+    float trim_yaw_heading_reading;
+//    Vector2f filt_att_fdbk_from_velxy_cd;
```
probably best to remove this commented out code..

bnsgeyer replied:
done

bnsgeyer commented:
> I guess at some point you'll need to rebase this on master because PR #18658 has renamed "inertial_nav::get_position();" to "inertial_nav->get_position_neu_cm();"

@rmackay9  completed.

bnsgeyer commented on ArduCopter/config.h:
```diff
@@ -55,7 +55,7 @@
   # define RC_FAST_SPEED                        125
   # define WP_YAW_BEHAVIOR_DEFAULT              WP_YAW_BEHAVIOR_LOOK_AHEAD
-  # define AUTOTUNE_ENABLED                     DISABLED
+  # define AUTOTUNE_ENABLED                     ENABLED
```
done.  Thanks!

bnsgeyer commented on libraries/AC_AutoTune/AC_AutoTune.cpp:
```diff
@@ -1713,56 +1249,585 @@ void AC_AutoTune::get_poshold_attitude(float &roll_cd_out, float &pitch_cd_out,
+        //        gcs().send_text(MAV_SEVERITY_INFO, "AutoTune: min_meas_cnt=%f", (double)(min_meas_cnt));
+    }
+
+    if (tgt_rate > max_target && new_target) {
```
done

bnsgeyer commented:
@rmackay9, @peterbarker and @lthall
I am complete with the major rework of the subclasses.  Have pulled as much out of the parent class into the subclass as possible to make it easier to review the code.  I took a few minutes to reconstruct the AutoTune.cpp file for multicopter.  What I did was pull all of the multicopter methods (referenced in the parent AutoTune.cpp) into the parent AutoTune.cpp file.  I cleaned up the formatting to minimize the changes that diff would catch.  I also removed the comments and #defines from the diff at the top.  So the attached file shows how the master AutoTune.cpp file has changed.  This gives a really good look at some flow changes I had to make in order to separate multi and heli.  I also provide the reconstructed autotune.cpp file for reference.
[autotune_cpp diff.txt](https://github.com/ArduPilot/ardupilot/files/7785027/autotune_cpp.diff.txt)

[AC_AutoTune-pr_cpp.txt](https://github.com/ArduPilot/ardupilot/files/7785031/AC_AutoTune-pr_cpp.txt)

bnsgeyer commented:
@rmackay9, @peterbarker and @lthall
I have ran autotune with master and with this PR branch to compare the results.  I ran the autotune 3 times to get some idea of the variation in the tuning.  I performed pitch and roll autotune with the Callisto model.  The yaw autotune never completed in master even after running for 20 min.  It appears that the Rate D up reaches the max tuning gains and is unable to achieve the desired response so it just keeps on testing even though it has reached a gain limit.  So I used the built-in quad model to test the yaw autotune.   The result compare very well and are statistically the same.  Here is the results for the average values, standard deviation and the % difference between master and the PR branch.
![image](https://user-images.githubusercontent.com/9907906/147606590-969d9582-0b6e-4777-abf3-27e39edf4c1c.png)
the largest difference is seen in the yaw autotune.  I'm pretty satisfied that the multicopter autotune has not been affected by the incorporation of heli autotune.

rmackay9 commented on libraries/AC_AutoTune/AC_AutoTune.cpp:
```diff
@@ -268,71 +162,18 @@ const char *AC_AutoTune::type_string() const
-    gcs().send_text(MAV_SEVERITY_INFO, "AutoTune: success %u/%u", counter, AUTOTUNE_SUCCESS_COUNT);
-
-    announce_time = now;
+    return "Bug";
```
Not a blocker but I think we should replace "Bug" with "Unknown" and if it should never happen then add a line like this to trigger an internal error.

```
// this should never happen
INTERNAL_ERROR(AP_InternalError::error_t::flow_of_control);
```

bnsgeyer replied:
Yeah not something I added but  I can make the change.  

Is that all you got ;)

rmackay9 replied:
@bnsgeyer i'm thinking about giving my IRIS an autotune with this and if it goes OK then I might merge it...

bnsgeyer replied:
Cool. I spoke with Leonard not long ago and he also offered to test it once you were happy with it. Either way works for me.  
what about the parameter issue that you needed to sort out.

bnsgeyer replied:
Incorporated requested change.

rmackay9 commented on libraries/AC_AutoTune/AC_AutoTune_Multi.cpp:
```diff
@@ -0,0 +1,1219 @@
+        gcs().send_text(MAV_SEVERITY_INFO, "AutoTune: lean=%f target=%f", (double)lean_angle, (double)target_angle);
+    }
+    if (!is_zero(rotation_rate)) {
+        gcs().send_text(MAV_SEVERITY_INFO, "AutoTune: rotation=%f target=%f", (double)(rotation_rate*0.01f), (double)(target_rate*0.01f));
```
Just checking the difference in output vs master and I see the, "rotation= xxx" message is new.  I wonder if this adds any value to the user.

rmackay9 replied:
sorry, I'm wrong, I see it now in master as well..

rmackay9 commented on libraries/AC_AutoTune/AC_AutoTune_Heli.h:
```diff
@@ -0,0 +1,285 @@
+    void set_gains_post_tune(AxisType test_axis) override;
+
+    // reverse direction for twitch test
+    bool twitch_reverse_direction() override { return positive_direction; }
```
This works but is a bit non-standard.  A more AP-ish way to do this would be to have this method return "true" for multicopters and "false" for tradheli's and then change this code in AC_AutoTune::control_attitude():

```
        // reverse direction for multicopter twitch test
        positive_direction = twitch_reverse_direction();
```

.. to be ..
```

        // reverse direction for multicopter twitch test
        if (twitch_reverse_direction()) {
            positive_direction = !positive_direction;
        }
```

I'm not suggesting we should change this now though but maybe someday..

rmackay9 commented:
Here are some results for multicopters and quadplanes (master vs this PR)
![multicopter-master-vs-new](https://user-images.githubusercontent.com/1498098/152077933-50a3ff8f-b130-4601-9fac-385762cefac0.png)

![quadplane-master-vs-new](https://user-images.githubusercontent.com/1498098/152077938-cf9f8883-adbf-4fa7-9a34-860e1377c69b.png)

rmackay9 commented:
I've also confirmed that the gains are returns to the original gains during pilot override and after switching to RTL

bnsgeyer commented:
@rmackay9 @tridge @lthall Thanks for all of your help in developing this capability!!!
Merging!
